### ISSUE TITLE
Usage of geotorch in Hypernetwork

### ISSUE THREAD

**Issue by JonathanSchmidt1** [Lezcano/geotorch#12]
I am trying to use the orthogonal constrain in  a hyernetwork. So effectively I have a Linear network that returns another module that uses a weight matrix that needs to be constrained. 

```python
class Functional(nn.Module):
    def __init__(self,
                 weights):
        ''' :param weights: Shape: (batch, out_ch, in_ch)
        '''
        super().__init__()
        self.weights = nn.Parameter(weights.squeeze())
    def forward(self, S,L):
        DD = torch.einsum('ij,jk->ik', S, self.weights)
        return torch.einsum('ij,ij,ij->', DD, DD, L)

class FunctionalModel(nn.Module):
      def __init__(self, weights):
          super(FunctionalModel, self).__init__()
          self.functional = Functional(weights)
          geotorch.orthogonal(self.functional, "weights")
      def forward(self, S,L):
          return self.functional(S,L)
```

The Hypernetwork returns an object of the FunctionalModel class, however when the geotorch constraint is included the Functionalmodel.functional.weights matrix changes each time the forward method of the hypernetwork is called with the same input (the constraint is however still fulfilled). Is there a way to work around this?
cheers and thank you very much.

**Comment by Lezcano** [comment]
I do not think I fully understand the problem here. Can you provide a minimal working example of this behaviour?

**Comment by JonathanSchmidt1** [comment]
Hope this example helps clear it up (tested with torch1.6):

import torch, geotorch, torch.nn as nn


class FCLayer(nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(in_features, out_features),
            nn.Tanh(),
        )

    def forward(self, input):
        return self.net(input)


class FCBlock(nn.Module):
    def __init__(self,
                 in_features,
                 out_features):
        super().__init__()

        self.net = FCLayer(
                in_features=in_features,
                out_features=out_features)

    def forward(self, input):
        return self.net(input)



class HyperLinear(nn.Module):
    '''A hypernetwork that predicts a single linear layer (weights).'''

    def __init__(self,
                 in_ch,
                 out_ch,
                 hyper_in_ch):

        super().__init__()
        self.in_ch = in_ch
        self.out_ch = out_ch

        self.hypo_params = FCBlock(in_features=hyper_in_ch,
                                   out_features=(in_ch * out_ch))

    def forward(self, hyper_input):
        hypo_params = self.hypo_params(hyper_input)

        # Indices explicit to catch erros in shape of output layer
        weights = hypo_params[..., :self.in_ch * self.out_ch]
        weights = weights.view(*(weights.size()[:-1]), self.out_ch, self.in_ch)
        return FunctionalModel(weights=weights)

# up to here simplified Hypernetwork definition (Fully connected layer that outputs the weights for one linear layer)

class Functional(nn.Module):
    def __init__(self,
                 weights):
        ''' layer whose weights should be constrained and are an output of the hypernetwork
        '''
        super().__init__()
        self.weights = nn.Parameter(weights.squeeze())
    def forward(self):
        #output should be the same as long as the input to the hypernetwork stays constant
        return torch.sum(self.weights)

#Module with constraint
class FunctionalModel(nn.Module):
      def __init__(self, weights):
          super(FunctionalModel, self).__init__()
          self.functional = Functional(weights)
          geotorch.orthogonal(self.functional, "weights")
      def forward(self):
          return self.functional()

        
size=5
model =  HyperLinear(in_ch=size,
                 out_ch=size,
                 hyper_in_ch=1)    

hyperinput = torch.tensor([[1.0]])
print('should be the same')
print(model(hyperinput).forward())
print(model(hyperinput).forward())




#same example without constraint
class FunctionalModel_without_constraint(nn.Module):
      def __init__(self, weights):
          super(FunctionalModel_without_constraint, self).__init__()
          self.functional = Functional(weights)
#          geotorch.orthogonal(self.functional, "weights")
      def forward(self):
          return self.functional()

class HyperLinear_without_constraint(nn.Module):
    '''A hypernetwork that predicts a single linear layer (weights).'''

    def __init__(self,
                 in_ch,
                 out_ch,
                 hyper_in_ch):

        super().__init__()
        self.in_ch = in_ch
        self.out_ch = out_ch

        self.hypo_params = FCBlock(in_features=hyper_in_ch,
                                   out_features=(in_ch * out_ch))
#        self.hypo_params[-1].apply(last_hyper_layer_init)

    def forward(self, hyper_input):
        hypo_params = self.hypo_params(hyper_input)
        weights = hypo_params[..., :self.in_ch * self.out_ch]
        weights = weights.view(*(weights.size()[:-1]), self.out_ch, self.in_ch)
        return FunctionalModel_without_constraint(weights=weights)

model =  HyperLinear_without_constraint(in_ch=size,
                 out_ch=size,
                 hyper_in_ch=1)    
hyperinput = torch.tensor([[1.0]])
print('Are the same without constraint')
print(model(hyperinput).forward())
print(model(hyperinput).forward())

**Comment by JonathanSchmidt1** [comment]
Hope this example explains what I am trying to do https://colab.research.google.com/drive/16hiNR8TO1M5uJvw_nTChGjWVtx6OsB3Y?usp=sharing

**Comment by Lezcano** [comment]
Could you make the colab public?

**Comment by JonathanSchmidt1** [comment]
Sorry, does it work for you now?

**Comment by Lezcano** [comment]
It does now, thanks. I'll take a look at it in a bit.

**Comment by JonathanSchmidt1** [comment]
Thought about the problem a bit more and it might be the best idea to just add the parameterization of the orthogonal matrix as part of the hyperparameter instead of doing as part of geotorch, (added it to the notebook). If you want you can close the Issue, Thank you for your quick help.
Ps:
Is the cayley or the matrix exponential representation more efficient?

**Comment by Lezcano** [comment]
Hi, sorry, I have not had time to look into this today.
I just had a look, and it turns out that what's happening is that every time that you put the constraint, the constraint is initialised. Therefore the randomness. If you put an orthogonal constraint on a linear layer `nn.Linear(3,3)`, the default weight from this layer is not orthogonal, and it is not clear which orthogonal weight it should be initialised to. As such, when you do `geotorch.orthogonal(layer, "weight")`, it initialises `layer.weight` to a randomly sampled orthogonal matrix. See:
https://geotorch.readthedocs.io/en/latest/orthogonal/stiefel.html#geotorch.Stiefel.sample 
The sampling happens here (see line 22)
https://github.com/Lezcano/geotorch/blob/55a91ca973ddeeb12d193d464af4b23da93d5ab4/geotorch/constraints.py#L18-L23
I hope this helps you understanding a bit better the inner workings of GeoTorch! I think I should make this a bit more explicit somewhere.

If you do not like this behaviour, you can always initialise your layer to whichever orthogonal matrix you like by doing
```python
layer = nn.Linear(3, 3)
geotorch.orthogonal(layer, "weight")
X = my_sampling_method(3, 3)
layer.weight = X
```

About the Cayley vs Exponential, the exponential map, as it is implemented in PyTorch, is more efficient than the Cayley map. That being said, I'd suggest you time them for the sizes that you are going to use with `timit` on a Jupyter notebook, just to be on the safe side.

**Comment by Lezcano** [comment]
As a separate minor point, I would recommend you to use `geotorch`'s methods over implementing them yourself, as they have a number of small optimisations and little quirks that you can benefit from at zero cost. It also has the advantage that it's more explicit with the intentions of what you want to do, but that's another story :)

**Comment by JonathanSchmidt1** [comment]
Thank you for your help, I thought that it was a problem like this with geotorch but I did not see how to fix it, now I should be able to bring it back into the code
